### PR TITLE
Introduce RoutePattern (#585)

### PR DESCRIPTION
**Pull request by rynowak**
* Introduce RoutePattern

Introduces RoutePattern - a new parser and representation for routing
templates, defaults, and constraints.

This is a new representation for all of the 'inputs' to routing that is
immutable and captures 'out of line' information for defaults and
constraints.

This will allow us to unify the handling of constraints and values from
attribute style routes and conventional style routes.

### CODE REVIEW

**Comment by rynowak** [comment]
Just a merge to master. Will merge when CI passes.

File: src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+            // This means that for parameters that have 'out of line' defaults we will modify
+            // the parameter to contain the default (same story for constraints).
+            //
+            // We also maintain a collection of defaults and constraints that will also
+            // contain the values that don't match a parameter.
+            //
+            // It's important that these two views of the data are consistent. We don't want
+            // values specified out of line to have a different behavior.
+
+            var updatedDefaults = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs]: Probably can't do a `ToLowerOrdinalIgnoreCase` :(

[the remark this answers is not in the document]
Comment (benaadams) [reply src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs]: Probably can't do a `ToLowerOrdinalIgnoreCase` :(

Might be easier to put a fix in dictionary

[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs]: Thanks Ben. None of this code is perf sensitive. This is used to parse the route template during startup. I could probably use your feedback over in https://github.com/aspnet/Routing/pull/591 though 😆

**Pull request merged**